### ISSUE TITLE
training step doesn't iterate over batch in order

### ISSUE THREAD

**Issue by kevinghst** [PyTorchLightning/pytorch-lightning#7125]
The lightning training steps don't seem to iterate over the batches in the dataloader in order.

How to reproduce:

So I initialize my train_dataloader without shuffling:

```python
def train_dataloader(self):
    return DataLoader(
        dataset=self.train_dataset,
        batch_size=self.args.batch_size,
        num_workers=self.args.num_workers,
        pin_memory=True,
        drop_last=True
    )
```

I then put a debugger inside training_step:
```python
def training_step(self, batch, batch_idx):
     pdb.set_trace()
     .....
```

To my frustration, the batch in the first break point does not correspond to the batch I get by iterating the dataloader manually as below:

`batch = next(iter(module.train_dataloader()))`

Why is that?

**Comment by carmocca** [comment]
Hi!

Keep in mind you must set the seed again every time the RNG is modified (any random number has been generated).
The following code shows that the batches match. Notice all the required `seed_everything` calls

```python
# No issues with this dataset as it has no randomness
class IndexDataset(Dataset):
    def __getitem__(self, index):
        return torch.tensor(index)

    def __len__(self):
        return 10

# Reproduces what you show if the `seed_everything` calls are removed
class RandomDataset(Dataset):
    def __getitem__(self, index):
        return torch.randn(1)

    def __len__(self):
        return 10


def test_bug(tmpdir):

    class TestModel(BoringModel):
        def training_step(self, batch, batch_idx):
            if batch_idx == 0:
                self.b0 = batch
            print(f'training_step {batch_idx}: {batch}')

        def train_dataloader(self):
            return DataLoader(RandomDataset(), shuffle=False)

    model = TestModel()
    model.val_dataloader = None
    model.training_epoch_end = None
    trainer = Trainer(default_root_dir=tmpdir, progress_bar_refresh_rate=0, fast_dev_run=3)
    tr_dl = model.train_dataloader()

    # Get the first batch
    pl.seed_everything(0)  # Seed the RNG state
    b0_0 = next(iter(tr_dl))

    # Again for sanity checking
    pl.seed_everything(0)  # Need this again as the RNG state has changed
    b0_1 = next(iter(tr_dl))
    print('Dataloader iter batch 0:', b0_0, b0_1)

    # Should be equal
    assert b0_0 == b0_1

    pl.seed_everything(0)  # Need this again as the RNG state has changed
    trainer.fit(model)

    # Should be equal
    assert b0_0 == model.b0
```

**Comment by Borda** [comment]
@kevinghst feel free to reopen if needed :]
